In [1]:
import pandas as pd 

### Read Train Dataset

In [2]:
df = pd.read_csv('train.csv')

This Dataset is different from Orginal Dataset in kaggle and I added some error in it manually. So first of all we need do some Data Engineering stuff to recognize and solve the issues

In [3]:
df.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0     1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0     1   
2                             Heikkinen, Miss. Laina  female  26.0     0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0     1   
4                           Allen, Mr. William Henry    male  35.0     0   

  Parch            Ticket     Fare Cabin Embarked  
0     0         A/5 21171   7.2500   NaN        S  
1     0          PC 17599  71.2833   C85        C  
2     0  STON/O2. 3101282   7.9250   NaN        S  
3     0            113803  53.1000  C123        S  
4     0            373450   8.0500   NaN        S

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          715 non-null    float64
 6   SibSp        891 non-null    object 
 7   Parch        891 non-null    object 
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(3), object(7)
memory usage: 83.7+ KB


In [5]:
df.describe()

PassengerId    Survived      Pclass         Age           Fare
count   891.000000  891.000000  891.000000  715.000000     891.000000
mean    446.000000    0.383838    2.353535   30.065972     607.347867
std     257.353842    0.486592    1.355652   17.457358   17170.623346
min       1.000000    0.000000    1.000000  -31.000000     -13.000000
25%     223.500000    0.000000    2.000000   20.000000       7.895800
50%     446.000000    0.000000    3.000000   28.000000      14.454200
75%     668.500000    1.000000    3.000000   38.000000      31.000000
max     891.000000    1.000000   33.000000  210.000000  512566.600000

### Short Description of features:
**Survived** = Passenger survived or not </br>
**PassengerId** = Passenger unique Id </br>
**Pclass** = Ticket class </br>
**Sex** = Passenger sex  </br>
**Age** = Passenger Age *in years*  </br>
**SibSp** = Number of siblings/spouses aboard the ship including the passenger  </br>
**ParCh** = Number of parents/children aboard the ship including the passenger  </br>
**Ticket** = Ticket number  </br>
**Fare** = Passenger fare  </br>
**Cabin** = Cabin number  </br>
**Embarked** = Port of passenger embarkation  </br>

The Dataset has 891 examples and 11 features + the target variable (Survived). 2 of the features are floats, 3 are integers and 7 are objects.
Let's take a look at each feature

In [6]:
df['Survived'].describe()

count    891.000000
mean       0.383838
std        0.486592
min        0.000000
25%        0.000000
50%        0.000000
75%        1.000000
max        1.000000
Name: Survived, dtype: float64

Everything seems fine

In [7]:
df['PassengerId'].describe()

count    891.000000
mean     446.000000
std      257.353842
min        1.000000
25%      223.500000
50%      446.000000
75%      668.500000
max      891.000000
Name: PassengerId, dtype: float64

Everything seems fine

In [8]:
df['Pclass'].describe()

count    891.000000
mean       2.353535
std        1.355652
min        1.000000
25%        2.000000
50%        3.000000
75%        3.000000
max       33.000000
Name: Pclass, dtype: float64

Something looks unnormal. max value and 75% and 50% and 25% and min does'n increase normal.let's look deeper on values

In [9]:
df['Pclass'].value_counts()

3     490
1     215
2     184
11      1
33      1
Name: Pclass, dtype: int64

Values 1,2,3 looks normal but we have one 11 and one 33 that based on feature description it seems these two values are error. so we should. let's look at their rows

In [10]:
df[df['Pclass']==11]

PassengerId  Survived  Pclass                      Name     Sex   Age  \
11           12         1      11  Bonnell, Miss. Elizabeth  Female  58.0   

   SibSp Parch  Ticket   Fare Cabin Embarked  
11     0     0  113783  26.55  C103        S

In [11]:
df[df['Pclass']==33]

PassengerId  Survived  Pclass             Name   Sex   Age SibSp Parch  \
401          402         0      33  Adams, Mr. John  male  26.0     0     0   

     Ticket  Fare Cabin Embarked  
401  341826  8.05   NaN        S

The other values seems alright and it seems there is no other issue in them. </br>
So we have two options: </br>
1- fill them with Null values </br>
2- based on the other values in this feature, maybe it was an input error so the value 11 is an error for value 1, and value 33 is an error for value 3. so maybe we can replace them with the currect values 1,3. </br>
I decided to replace them with the currect values.

In [12]:
df['Pclass'].replace(11,1, inplace=True)

In [13]:
df['Pclass'].replace(33,3, inplace=True)

In [14]:
df['Pclass'].value_counts()

3    491
1    216
2    184
Name: Pclass, dtype: int64

In [15]:
df['Pclass'].describe()

count    891.000000
mean       2.308642
std        0.836071
min        1.000000
25%        2.000000
50%        3.000000
75%        3.000000
max        3.000000
Name: Pclass, dtype: float64

now everything looks normal </br>
Let's go to the next feature

In [16]:
df['Sex'].describe()

count      891
unique      11
top       male
freq       571
Name: Sex, dtype: object

We have 8 unique values in Sex? it's not normal. Let's look inside

In [17]:
df['Sex'].unique()

array(['male', 'female', 'Female', 'malee', 'FEMALE', 'MALE', 'Male',
       'malle', 'femal', 'femmale', 'mmale'], dtype=object)

Oh, there is some Upper case and Lower case version of Male and Female value.

In [18]:
df['Sex'].value_counts()

male       571
female     310
malle        2
Female       1
malee        1
FEMALE       1
MALE         1
Male         1
femal        1
femmale      1
mmale        1
Name: Sex, dtype: int64

Ok, male and female are main values and the others should be convert to lower case

In [19]:
df['Sex'] = df['Sex'].apply(str.lower)

In [20]:
df['Sex'].value_counts()

male       573
female     312
malle        2
malee        1
femal        1
femmale      1
mmale        1
Name: Sex, dtype: int64

Ok, besides that, we have some other wrong input for male like malle and for female like femal. Let's correct them too.

In [21]:
female_group = ['femal', 'femmale']
df['Sex'].replace(female_group, 'female', inplace=True)

In [22]:
male_group = ['malle', 'malee', 'mmale']
df['Sex'].replace(male_group, 'male', inplace=True)

In [23]:
df['Sex'].value_counts()

male      577
female    314
Name: Sex, dtype: int64

now everything looks normal </br>
Let's go to the next feature

In [24]:
df['Age'].describe()

count    715.000000
mean      30.065972
std       17.457358
min      -31.000000
25%       20.000000
50%       28.000000
75%       38.000000
max      210.000000
Name: Age, dtype: float64

Wow, min and max values seems incorrect. let's check them all

In [25]:
df[(df['Age'] < 0) | (df['Age'] > 100)]

PassengerId  Survived  Pclass  \
12            13         0       3   
174          175         0       1   
237          238         1       2   
368          369         1       3   
432          433         1       2   
673          674         1       2   

                                                  Name     Sex    Age SibSp  \
12                      Saundercock, Mr. William Henry    male  -20.0     0   
174                            Smith, Mr. James Clinch    male  156.0     0   
237                   Collyer, Miss. Marjorie "Lottie"  female   -8.0     0   
368                                Jermyn, Miss. Annie  female  210.0     0   
432  Louch, Mrs. Charles Alexander (Alice Adelaide ...  female  142.0     1   
673                              Wilhelms, Mr. Charles    male  -31.0     0   

    Parch      Ticket     Fare Cabin Embarked  
12      0   A/5. 2151   8.0500   NaN        S  
174     0       17764  30.6958    A7        C  
237     2  C.A. 31921  26.2500   NaN        S  
368     0       14313   7.7500   NaN        Q  
432     0  SC/AH 3085  26.0000   NaN        S  
673     0      244270  13.0000   NaN        S

For negetive values for Age we can replace them with exact positive values and also we can replace them with Null values. For values more than usual maximum age, let's take a look at higher values

In [26]:
df[(df['Age'] >= 0) & (df['Age'] <= 100)]['Age'].max()

80.0

So, the maximum Age under 100 is 80, that looks normal for a passenger. Let's continue remove incorrect Age values

In [27]:
import numpy as np

In [28]:
df.loc[(df['Age'] < 0) | (df['Age'] > 100), 'Age'] = np.nan

In [29]:
df['Age'].describe()

count    709.000000
mean      29.687123
std       14.509245
min        0.420000
25%       20.500000
50%       28.000000
75%       38.000000
max       80.000000
Name: Age, dtype: float64

now everything looks normal </br>
Let's go to the next feature

In [37]:
df['SibSp'].describe()

count    891.000000
mean       0.517396
std        1.105899
min       -2.000000
25%        0.000000
50%        0.000000
75%        1.000000
max        8.000000
Name: SibSp, dtype: float64

SibSp comes for Number of First-degree relatives boarded on ship together but type of this feature is Object so something is wrong. lets look at unique values

In [31]:
df['SibSp'].unique()

array(['1', '0', '3', '4', 'NO', '2', '5', '-1', '8', '-2', 'YES'],
      dtype=object)

We have some incorrect enteries here, like 'NO' or 'YES' or even -1 and -2. lets count them all

In [40]:
df['SibSp'].value_counts()

0    607
1    210
2     28
4     18
3     16
8      7
5      5
Name: SibSp, dtype: int64

Ok, there is a few incorrect values. Let's check them one by one: </br>
***NO*** means the passenger is alone so we can replace them with 0 </br>
***YES*** means the passenger is not alone but we don't know how many other passengers are with this passenger so we can't replace it with a exact number. The count of values says that 0 with 605 frequency is on top of the list, but we are looking for a number except 0. the next one is 1 with 207 frequency, so we can replace 'YES' with 1.</br>
***-1 and -2*** could be a entery mistake more than a incorrect value, so we can simply positive them

In [33]:
df['SibSp'].replace('NO', 0, inplace=True)

In [34]:
df['SibSp'].replace('YES', 1, inplace=True)

Now, the feature has not any non-number values so before processing ***-1*** and ***-2*** it's better to convert the feature type to int

In [35]:
df['SibSp'] = df['SibSp'].astype(int)

In [39]:
def make_positive(x):
    if x < 0:
        return x * -1
    return x

df['SibSp'] = df['SibSp'].apply(make_positive)

In [41]:
df['SibSp'].unique()

array([1, 0, 3, 4, 2, 5, 8], dtype=int64)

In [42]:
df['SibSp'].value_counts()

0    607
1    210
2     28
4     18
3     16
8      7
5      5
Name: SibSp, dtype: int64

now everything looks normal </br>
Let's go to the next feature

In [45]:
df['Parch'].describe()

count     891
unique     11
top         0
freq      674
Name: Parch, dtype: object

In [46]:
df['Parch'].unique()

array(['0', '1', '2', '5', 'O', '3', '4', 'PC', 'F.C.C', 'WE', '6'],
      dtype=object)

In [47]:
df['Parch'].value_counts()

0        674
1        117
2         79
5          5
3          5
4          4
O          3
PC         1
F.C.C      1
WE         1
6          1
Name: Parch, dtype: int64

The Parch feature is also about number of First-Class family so it should be number but its type is object because there is some incorrect values like ***PC*** or ***F.C.C*** or ***O*** in it. </br>
***O*** and ***0*** are the same in many cases so we can replace ***O*** with ***0*** </br>
the other String values are incorrect and should replace with ***0***</br>

In [50]:
def convert_to_int(x):
    try:
        return int(x)
    except ValueError as ve:
        return 0
    

df['Parch'] = df['Parch'].apply(convert_to_int)

because in every scenario the incorrect value should replace with ***0*** and type of all incorrect values is *string*, so I try to convert every value in this feature to int and ofcourse string values raise error during conversion so simply the function return ***0*** in exceptions.

In [51]:
df['Parch'].value_counts()

0    680
1    117
2     79
5      5
3      5
4      4
6      1
Name: Parch, dtype: int64

now everything looks normal </br>
Let's go to the next feature